### PR TITLE
Persist license key and support configuration update

### PR DESCRIPTION
**Pull request by ramizdundar**
License key changes now can be persisted. Also users can set their licenses via configuration update & reloading.

EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/4579

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

**Comment by ramizdundar** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/internal/config/AbstractConfigLocator.java [hunk 1]
```diff
@@ -113,6 +115,11 @@ public boolean isConfigPresent() {
      */
     public abstract boolean locateDefault();
 
+    public static boolean isDefaultConfigFile(File file) {
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/config/AbstractConfigLocator.java:118]: I think this is not a good criterion for checking if `file` is a default config file.
Assuming I have a file named `/opt/hazelcast/hazelcast-default.xml` and use it to start hazelcast with `-Dhazelcast.config=/opt/hazelcast/hazelcast-default.xml`, should the above scenario fail? I assume it's loaded from the "default" file if it was loaded by `AbstractConfigLocator#loadDefaultConfigurationFromClasspath`, not if just the filename matches.

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java [hunk 2]
```diff
@@ -205,6 +205,14 @@ public void broadcastConfig(IdentifiedDataSerializable config) {
         future.joinInternal();
     }
 
+    @Override
+    public void updateLicense(String licenseKey) {
+        // NOP
+
+        // Maybe we can throw exception here in the future. For now, I'm
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java:212]: Isn't throwing an exception clearer?

File: hazelcast/src/test/java/com/hazelcast/internal/config/dynamic/DynamicConfigXmlGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        StringBuilder xmlBuilder = new StringBuilder();
+        ConfigXmlGenerator.XmlGenerator gen = new ConfigXmlGenerator.XmlGenerator(xmlBuilder);
+
+        // generate string
+        xmlBuilder.append("<hazelcast ")
+                .append("xmlns=\"http://www.hazelcast.com/schema/config\"\n")
+                .append("xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\n")
+                .append("xsi:schemaLocation=\"http://www.hazelcast.com/schema/config ")
+                .append("http://www.hazelcast.com/schema/config/hazelcast-config-")
+                .append(Versions.CURRENT_CLUSTER_VERSION.getMajor())
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/internal/config/dynamic/DynamicConfigXmlGeneratorTest.java:57]: minor: could use `Versions.CURRENT_CLUSTER_VERSION.toString()` which returns the major.minor string

**Review by vbekiaris — CHANGES_REQUESTED** [review]

File: hazelcast/src/test/java/com/hazelcast/internal/config/dynamic/DynamicConfigXmlGeneratorTest.java (hunk 3 above)
Comment (ramizdundar) [reply hazelcast/src/test/java/com/hazelcast/internal/config/dynamic/DynamicConfigXmlGeneratorTest.java]: https://github.com/hazelcast/hazelcast/pull/20446/commits/b687eefd6951adb6d998756cac0020629d33e5f3

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java (hunk 2 above)
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java]: https://github.com/hazelcast/hazelcast/pull/20446/commits/d8c9908483cc67bbeddead8a040f40886fbe4d67

File: hazelcast/src/main/java/com/hazelcast/internal/config/AbstractConfigLocator.java (hunk 1 above)
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/internal/config/AbstractConfigLocator.java]: https://github.com/hazelcast/hazelcast/pull/20446/commits/87e10baddea47859440058d7dc87a55d75d00195

File: hazelcast/src/main/java/com/hazelcast/internal/config/LicenseKey.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    private final String licenseKey;
+
+    public LicenseKey(String licenseKey) {
+        this.licenseKey = licenseKey;
+    }
+
+    public String getLicenseKey() {
+        return licenseKey;
+    }
+}
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/internal/config/LicenseKey.java:29]: maybe equals / hashcode / toString (without the actual key)?

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: hazelcast/src/main/java/com/hazelcast/config/Config.java [hunk 5]
```diff
@@ -334,17 +335,23 @@ public static Config loadFromClasspath(ClassLoader classLoader, String resource,
         checkTrue(resource != null, "resource can't be null");
         checkTrue(properties != null, "properties can't be null");
 
-        InputStream stream = classLoader.getResourceAsStream(resource);
+        InputStream stream;
+        URL url = classLoader.getResource(resource);
+        try {
+            stream = url != null ? url.openStream() : null;
+        } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/config/Config.java:342]: This exception was also swallowed previously. Please check getResourceAsStream() function. I inheridted it's behaviour. But I still can log it if you prefer it.

[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast/src/main/java/com/hazelcast/config/Config.java:342]: I see. Could you please add a comment above this block that you basically inlined the `getResourceAsStream` so you could also use the url.

[the remark this answers is not in the document]
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/config/Config.java:346]: https://github.com/hazelcast/hazelcast/pull/20446/commits/bbfc4a3949421554daf689852db64b9f5fb0d5c9

File: hazelcast/src/main/java/com/hazelcast/internal/config/LicenseKey.java (hunk 4 above)
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/internal/config/LicenseKey.java:59]: https://github.com/hazelcast/hazelcast/pull/20446/commits/66601424b00ed564636c7a9733a799e8099c453b

**Review by frant-hartm — APPROVED** [review]

**Review by vbekiaris — APPROVED** [review]
+1, a minor suggestion

File: hazelcast/src/main/java/com/hazelcast/internal/config/LicenseKey.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    @Override
+    public int hashCode() {
+        return licenseKey.hashCode();
+    }
+
+    @Override
+    @SuppressWarnings("checkstyle:MagicNumber")
+    public String toString() {
+        // last 8 characters are visible
+        String maskedLicense = MASK_FOR_SENSITIVE_DATA + licenseKey.substring(licenseKey.length() - 8);
```
[the remark this answers is not in the document]
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/internal/config/LicenseKey.java]: https://github.com/hazelcast/hazelcast/pull/20446/commits/fe13f95d74eb5f479754c1062d8cc0b1a6cb1e10

**Pull request merged**